src/test/scala/vectorpipe/vectortile/TestPipeline.scala

```diff
@@ -0,0 +1,55 @@
+package vectorpipe.vectortile
+
+import geotrellis.raster.RasterExtent
+import geotrellis.spark.SpatialKey
+import geotrellis.spark.tiling._
+import geotrellis.vector._
+import geotrellis.vectortile._
+
+import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema
+import org.apache.spark.sql.functions
+import org.apache.spark.sql.functions.{array, col, explode, lit, sum}
+import org.apache.spark.sql.types._
+import org.locationtech.jts.{geom => jts}
+
+import vectorpipe._
+import vectorpipe.vectortile._
+
+case class Bin(x: Int, y: Int)
+object Bin {
+  def apply(tup: (Int, Int)): Bin = Bin(tup._1, tup._2)
+}
+
+case class TestPipeline(geometryColumn: String, baseOutputURI: java.net.URI, gridResolution: Int) extends Pipeline {
+  val weightedCentroid = new WeightedCentroid
+
+  override def reduce(input: DataFrame, layoutLevel: LayoutLevel, keyColumn: String): DataFrame = {
+    import input.sparkSession.implicits._
+
+    val layout = layoutLevel.layout
+    val binOfTile = functions.udf { (g: jts.Geometry, key: GenericRowWithSchema) =>
+      val pt = g.asInstanceOf[jts.Point]
+      val k = SpatialKey(key.getInt(0), key.getInt(1))
```

Comment:
`key` is a tuple of `(x, y)` rather than a serialized `SpatialKey`?

Reply:
I think I need an Encoder/Decoder pair for `SpatialKey`, otherwise these are returned to me as a `GenericRowWithSchema`.  I tried to `getAs[SpatialKey]`, and it told me to take a hike.  Any thoughts?

Reply:
With `SparkSession.implicits._` in scope, it's worked for me in the past (see `EditHistogram` in OSMesa).

Reply:
Maybe that's a `Dataset` vs. `DataFrame` thing though...

Reply:
I can tell you that I definitely tried the other way.  Want a solution to this, as it's very cumbersome.

Reply:
I think you're thinking of https://stackoverflow.com/questions/39833185/convert-dataframe-back-to-rdd-of-case-class-in-spark ... that seems to be a dataframe-specific thing?

Reply:
Maybe?

This is how OSMesa does it: https://github.com/azavea/osmesa/pull/110/files#diff-6679397ac12868f224502e17764fda75R27

It's a member of a case class, so `ExpressionEncoder` handles it (and there's apparently nothing weird that requires a custom encoder for `SpatialKey`).

What did it tell you while telling you to take a hike?

Reply:
Something along the lines of not being able to cast a WrappedArray to SpatialKey.  I could handle this by putting an implicit in scope.  I wonder if there is a provided implicit that I haven't imported?